lib/StackAnalysis/StackAnalysis.cpp

```diff
@@ -383,6 +434,1140 @@ void commitToModel(GeneratedCodeBasicInfo &GCBI,
+                           const std::set<GlobalVariable *> &WrittenRegisters,
+                           SortedVector<model::BasicBlock> &CFG,
+                           llvm::Function *F) {
+  using namespace llvm;
+
+  SmallVector<std::pair<CallInst *, uint64_t>, 4> MaybeReturns;
+  SmallSet<std::pair<CallInst *, FunctionEdgeTypeValue>, 4> Result;
+  std::set<GlobalVariable *> CalleeSavedRegs;
+  std::set<GlobalVariable *> ClobberedRegs(ABIRegisters.begin(),
+                                           ABIRegisters.end());
+
+  for (CallBase *Call : callers(IndirectBranchInfoMarker)) {
+    auto *CI = cast<CallInst>(Call);
+    if (CI->getParent()->getParent() == F) {
+      if (isa<ConstantInt>(CI->getArgOperand(0))) {
+        if (isa<ConstantInt>(CI->getArgOperand(1))) {
+          int64_t FSO = cast<ConstantInt>(CI->getArgOperand(1))->getSExtValue();
+          if (FSO >= 0)
+            MaybeReturns.emplace_back(CI, FSO);
+          else
+            Result.insert({ CI, FunctionEdgeTypeValue::BrokenReturn });
+        } else {
+          // TODO: should this be a longjmp or a brokenreturn?
+          Result.insert({ CI, FunctionEdgeTypeValue::LongJmp });
+        }
+      }
+    }
+  }
+
+  // Elect a final stack offset.
+  const auto WinFSO = electFSO(MaybeReturns);
+
+  // Did we find a valid return instruction?
+  for (const auto &E : MaybeReturns) {
+    auto *CI = E.first;
+    if (E.second == *WinFSO) {
+      Result.insert({ E.first, FunctionEdgeTypeValue::Return });
+      unsigned NumArgs = CI->getNumArgOperands();
+      if (NumArgs > 3) {
+        for (unsigned Idx = 3; Idx < NumArgs; ++Idx) {
+          if (isa<ConstantInt>(CI->getArgOperand(Idx))) {
+            if (cast<ConstantInt>(CI->getArgOperand(Idx))->getZExtValue() == 0)
+              CalleeSavedRegs.insert(ABIRegisters[Idx - 3]);
+          }
+        }
+      }
+    } else {
+      Result.insert({ E.first, FunctionEdgeTypeValue::BrokenReturn });
+    }
+  }
+
+  // Neither a return nor a broken return?
+  for (CallBase *Call : callers(IndirectBranchInfoMarker)) {
+    auto *CI = cast<CallInst>(Call);
+    if (CI->getParent()->getParent() == F) {
+      if (isa<ConstantInt>(CI->getArgOperand(0)))
+        continue;
+
+      auto *V = dyn_cast<ConstantInt>(CI->getArgOperand(1));
+      if (WinFSO.has_value() && V != nullptr && V->getZExtValue() == *WinFSO)
+        Result.insert({ CI, FunctionEdgeTypeValue::IndirectTailCall });
+      else
+        Result.insert({ CI, FunctionEdgeTypeValue::LongJmp });
+    }
+  }
+
+  bool FoundRet, FoundBrokenRet;
+  FoundRet = FoundBrokenRet = false;
+  for (const auto &E : Result) {
+    if (E.second == FunctionEdgeTypeValue::Return)
+      FoundRet = true;
+    if (E.second == FunctionEdgeTypeValue::BrokenReturn)
+      FoundBrokenRet = true;
+  }
+
+  FunctionTypeValue Type;
+  if (FoundRet) {
+    Type = FunctionTypeValue::Regular;
+  } else if (FoundBrokenRet) {
+    Type = FunctionTypeValue::Fake;
```

Comment:
Didn't we say that we want to mark as fake if there is one *and only one* broken return?

Reply:
Which is what is happening right now, no?

Reply:
No, here you're marking it as fake if there's *at least* one broken return.

Reply:
Uhmm, and doesn't it sound reasonable to mark as fake if, for instance, all returns are broken? What would it be otherwise?

Reply:
If there are multiple returns, there might be different stack heights.
Also, I think we said we wanted to have fake function only if we the stack offset is known.
I don't remember exactly why, but at a certain point we said "we need to significantly restrict the situations in which we mark as fake".
Do you remember?

Reply:
I do. Indeed, as of now, we say a return may be broken only if the stack offset is known and less then zero (e.g., `-8`). I'm wondering how should fix this then, to take into account the case we have _one and only one_ broken return. I guess we should count the number of `indirect_branch_info` we have?

Reply:
I'd like to remember why we decided to be so strict... I don't want to put arbitrary restrictions of unknown reason.
It was something recent when you were testing inlined fake function...

Reply:
This should be fixed.